Nerfing Unions and Coalitions

FelModder opened the pull request:
<!--
A basic changelog, goes to patch notes of the next version, so should be as short and informative as possible.
-->
## Changelog:
- Reduced the amount of possible Union members from 30+ to 10+ only
- Those characters affected by Isolation mechanics (Kingdom of Suramar, all of Pandaria, all of Uldum) cannot join coalitions against crisis they are isolated from
- Cannot join a Coalition if you are already in a Union
- Union requirements actually work as they are intended: only certain religions and cultures can join their canonical Unions
- Increased chance of characters leaving your Union if there are no Crisis around in your continent

<!--
A changelog where you can describe more complicated things to other developers and testers.
-->
## Developer changelog:
- Fixed an issue where you couldn't call your allies to war if you are at war with a Crisis entity
- LOTS of error log fixes, hence increasing the performance of the game
- Fixed an LOC error where `Nomadic Horde` showed up in the crisis list instead of the Orcish Horde

<!--
Before the merge, we recommend you to do these tests with your branch.
-->
## Tests:
- [ ] There are no errors in `wc` files in `Documents\Paradox Interactive\Crusader Kings III\logs\error.log` except `portrait_decals.cpp:101`
- [ ] The mod takes less than 5.5 GB in the Task Manager (Windows)

<!--
If you need to explain something to testers. Otherwise, delete this part.
-->
# How to test:

FelModder committed:
reducing amount for union_overcrowded_trigger and reducing factors for joining

FelModder committed:
has_alliance_race_trigger for Alliance Limitations

FelModder committed:
has_horde_race_trigger for Horde Limitations

FelModder committed:
has_illidari_race_triggers for Illidari Limitations

FelModder committed:
has_draconic_race_trigger for Draconic Limitations

FelModder committed:
has_scourge_race_trigger for Scourge Limitations

FelModder committed:
has_axisawful_race_trigger for AWFUL RACES

Grekopithikos commented on common/scripted_triggers/wc_pol_faction_triggers.txt:
```diff
@@ -678,6 +703,76 @@ has_alliance_culture_trigger = {
+		has_trait_with_flag = dragon_family
+		has_trait = creature_goblin
+		has_trait = creature_pygmy
+		has_trait = creature_hozen
+		has_trait = creature_flamewaker
+		has_trait = creature_vrykul
+		has_trait = creature_tuskarr
+		has_trait = creature_magnataur
+		has_trait = creature_gorloc
+		has_trait = creature_wolvar
+		has_trait = creature_taunka
+		has_trait = creature_kvaldir
+		has_trait = creature_nerubian
+		has_trait = creature_harpy
+		has_trait = creature_mechagnome
+		has_trait = creature_giant
+		has_trait = creature_tauren
+		has_trait = creature_quilboar
+		has_trait = creature_tolvir
+		has_trait = creature_highborne
+		has_trait = creature_vulpera
+		has_trait = creature_sethrak
+		has_trait = creature_pandaren
+		has_trait = creature_mogu
+		has_trait = creature_saurok
+		has_trait = creature_yaungol
+		has_trait = creature_mantid
+		has_trait = creature_grummle
+		has_trait = creature_nightborne
+		has_trait = creature_drogbar
+		has_trait = creature_highmountain_tauren
+		has_trait = creature_trogg
+		has_trait = creature_faldorei
+		has_trait = creature_sea_giant
+		has_trait = creature_makrura
+		has_trait = creature_gilblin
+		has_trait = creature_tortollan
+		has_trait = creature_broken
+		has_trait = creature_lost_one
+	}
```
I don't think these parts are necessary

FelModder replied:
The fact the the Cultural/Religion triggers didn't use NOR and make exceptions is the reason why random faiths were able to join factions they weren't supposed to

So yes, this is necessary.

Grekopithikos replied:
Wouldn't this be fixed by putting something like
```
modifier = {
	factor = -2
	has_alliance_race_trigger = no
}
```
would that work?

FelModder replied:
that just means the races listed in OR = {} will not join either...

FelModder replied:
These `race_triggers` are here to not only avail the Canonical races to their respective Unions and properly balance out the chances of people joining the factions but to also restrict unrelated/un-historical races from joining what they aren't supposed to.

Union get too strong sometimes and restrict fun because they become overpowered so easily.

FelModder replied:
wait ok i found an alternative

fixed

Grekopithikos commented on common/scripted_triggers/wc_pol_faction_triggers.txt:
```diff
@@ -733,12 +828,85 @@ has_horde_culture_trigger = {
+		has_trait = creature_taunka
+		has_trait = creature_troll
+		has_trait = creature_nightborne
+		has_trait = creature_red_dragon
```
Are you sure about red dragons? They were made part of the Horde by force.

Also High Elves should be there because Blood Elves.

FelModder replied:
fixed

Grekopithikos commented on common/scripted_triggers/wc_pol_faction_triggers.txt:
```diff
@@ -812,6 +980,87 @@ has_illidari_culture_trigger = {
+has_illidari_race_trigger = {
+	OR = {
+		has_trait = demon_hunter
+		has_trait = creature_dryad
+		has_trait = creature_frostnymph
```
when were those part of the Illidari?

FelModder replied:
idk, i just put them in there because their cultural_heritage is in the cultures list for illidari cultures

Shall i remove both the trait and the cultures of those?

Grekopithikos replied:
I mean lorewise there is no precedent for them being there.
But if there are situations were a dryad is illidari religion then I can definitely see them joining.

Grekopithikos commented on common/scripted_triggers/wc_pol_faction_triggers.txt:
```diff
@@ -824,13 +1073,93 @@ has_draconic_culture_trigger = {
+		has_trait = creature_gorloc
+		has_trait = creature_watcher
+		has_trait = creature_taunka
+		has_trait = creature_wolvar
+		has_trait = creature_kvaldir
+		has_trait = creature_nerubian
+		has_trait = creature_harpy
+		has_trait = creature_mechagnome
+		has_trait = creature_giant
+		has_trait = creature_night_elf
+		has_trait = creature_tauren
+		has_trait = creature_centaur
+		has_trait = creature_quilboar
+		has_trait = creature_tolvir
+		has_trait = creature_qiraji
+		has_trait = creature_frostnymph
+		has_trait = creature_dryad
+		has_trait = creature_highborne
+		has_trait = creature_vulpera
+		has_trait = creature_sethrak
+		has_trait = creature_pandaren
+		has_trait = creature_mogu
+		has_trait = creature_jinyu
+		has_trait = creature_saurok
+		has_trait = creature_ankoan
+		has_trait = creature_yaungol
+		has_trait = creature_mantid
+		has_trait = creature_grummle
+		has_trait = creature_nightborne
+		has_trait = creature_drogbar
+		has_trait = creature_highmountain_tauren
+		has_trait = creature_trogg
+		has_trait = creature_faldorei
+		has_trait = creature_sea_giant
+		has_trait = creature_makrura
+		has_trait = creature_gilblin
+		has_trait = creature_tortollan
+		has_trait = creature_broken
+		has_trait = creature_lost_one
+	}
```
Once again these parts aren't necessary at all.
Race traits are mutually exclusive.

FelModder committed:
Use all these unused modifiers

Why weren't these used before?

FelModder committed:
Horde: Remove Red Dragons and put in High Elves

FelModder committed:
remove dryads and nymphs from illidari

FelModder committed:
okay i found an alternative

Grekopithikos commented on common/scripted_triggers/wc_pol_faction_triggers.txt:
```diff
@@ -805,13 +905,31 @@ has_custom_union_religion_trigger = {
+		has_trait = demon_hunter
+		has_trait = creature_high_elf
+		has_trait = creature_night_elf
+		has_trait = creature_highborne
+		has_trait = creature_nightborne
```
Highborne and Nightborne also don't have any history with the Illidari

FelModder replied:
fixed

FelModder committed:
fix Nerzhul becoming ancestral faith instead of death_god...

FelModder committed:
remove highborne and nightborne from illidari

FelModder committed:
add custom tooltips

FelModder committed:
this is better

Grekopithikos commented on common/scripted_triggers/wc_pol_faction_triggers.txt:
```diff
@@ -678,6 +790,28 @@ has_alliance_culture_trigger = {
+		has_trait = creature_high_elf
+		has_trait = creature_human
+		has_trait = creature_night_elf
+		has_trait = creature_watcher
```
Since when were watchers in the Alliance?

Grekopithikos commented on common/scripted_triggers/wc_pol_faction_triggers.txt:
```diff
@@ -678,6 +790,28 @@ has_alliance_culture_trigger = {
+	NOR = { # No Evil mfs allowed
+		has_trait = being_undead
+		has_trait = being_demon
+		has_trait = being_void
```
What about void elves?

FelModder replied:
Oh yea, completely forgot about them

Grekopithikos commented on common/scripted_triggers/wc_pol_faction_triggers.txt:
```diff
@@ -733,12 +867,32 @@ has_horde_culture_trigger = {
+		has_trait = creature_taunka
+		has_trait = creature_troll
+		has_trait = creature_nightborne
+		has_trait = creature_high_elf
```
Maybe put a
```
AND = {
has_trait = creature_human
has_trait = being_undead
}
```
there to represent the Forsaken?

FelModder replied:
I don't think this is necessary since undead are not restricted from joining the Horde at all.

FelModder replied:
ohh wait i get what you mean

Forsaken also have human undead, forgot about those

fixed

edmund-h reviewed (changes requested):
please make @Grekopithikos requested changes. 

also, did you actually restrict unions to these races or just add negative modifiers to join acceptance/ai request chance? because your localization sounds like it's completely restricted.

your overfocus on race as a qualifier is very shortsighted and could lead to some really stupid unions. you need to take more factors into account. honestly i think faith, not race, should be the deciding factor, but we should at least weigh both

Grekopithikos commented on common/scripted_triggers/wc_pol_faction_triggers.txt:
```diff
@@ -678,6 +790,28 @@ has_alliance_culture_trigger = {
+		has_trait = creature_draenei
+		has_trait = creature_dryad
+		has_trait = creature_frostnymph
+		has_trait = creature_centaur
```
Centaurs have never been part of the Alliance.

Grekopithikos commented on common/scripted_triggers/wc_pol_faction_triggers.txt:
```diff
@@ -678,6 +790,28 @@ has_alliance_culture_trigger = {
 }
+has_alliance_race_trigger = {
+	OR = {
+		has_trait = creature_draenei
```
should add Broken too since they've been in the Alliance

FelModder committed:
remove watchers from alliance

FelModder committed:
Void elves are a part of the Alliance

FelModder committed:
Undead Humans are Part of Forsaken too

FelModder committed:
Centaurs go solo

FelModder committed:
Broken can join the alliance as well

FelModder commented:
> please make @Grekopithikos requested changes.
> 
> also, did you actually restrict unions to these races or just add negative modifiers to join acceptance/ai request chance? because your localization sounds like it's completely restricted.
> 
> your overfocus on race as a qualifier is very shortsighted and could lead to some really stupid unions. you need to take more factors into account. honestly i think faith, not race, should be the deciding factor, but we should at least weigh both

I did all the changes as requested.

I did not overfocus on races, i just added them.

Faith, Culture , and now, Race are all deciding factors on which AI can join and which cannot.

The Faith and Culture restrictions were there before but not properly implemented inside the `can_join_union_trigger` hence it cause any random faith and culture to be able to join random unions.

Race trigger was also added because i fixed the unhinged factors and multipliers that made unions bloated an overpowered.

This race trigger is to not only restrict to certain race but to also balance out ai_chances modifiers for the AI joining a union.

FelModder committed:
Merge branch 'dev' into fix/code/rebalancing-and-nerfing-for-unions

FelModder committed:
Allow darkfallen elves to join Horde

Grekopithikos commented on common/scripted_triggers/wc_pol_faction_triggers.txt:
```diff
@@ -733,12 +865,39 @@ has_horde_culture_trigger = {
+	AND = {
+		has_trait = creature_high_elf
+		has_trait = being_undead
+	}
```
What about Blood Elves?

FelModder committed:
Blood elves for Horde and Illidari

FelModder committed:
remove race restrictions

FelModder committed:
I hate LOC ughh...

FelModder committed:
make it so that union_overcrowded_trigger is checked first

FelModder committed:
Higher chance of leaving if Union is overcrowded

FelModder committed:
Grim religions don't join except scourge

FelModder committed:
Merge branch 'dev' into fix/code/rebalancing-and-nerfing-for-unions

FelModder committed:
Fixing bad horde LOC

changed horde -> horde_invaders

FelModder committed:
fixes?

FelModder committed:
crises error log spam fix

FelModder committed:
cannot join coalition if you are already in union or are isolated

FelModder committed:
Custom tt if you are already in a union

FelModder committed:
cannot call ally against a crisis fix

FelModder committed:
error log fix

FelModder committed:
okay i'm done with fixes

FelModder committed:
Corruptism can join the horde
